src/python/pants/backend/python/rules/pex.py

```diff
@@ -4,23 +4,30 @@
 from pants.backend.python.rules.download_pex_bin import DownloadedPexBin
 from pants.backend.python.subsystems.python_native_code import PexBuildEnvironment, PythonNativeCode
 from pants.backend.python.subsystems.python_setup import PythonSetup
-from pants.engine.fs import Digest
+from pants.engine.fs import EMPTY_DIRECTORY_DIGEST, Digest, DirectoriesToMerge
 from pants.engine.isolated_process import (ExecuteProcessRequest, ExecuteProcessResult,
                                            MultiPlatformExecuteProcessRequest)
 from pants.engine.platform import Platform, PlatformConstraint
 from pants.engine.rules import optionable_rule, rule
 from pants.engine.selectors import Get
-from pants.util.objects import datatype, hashable_string_list, string_optional, string_type
+from pants.util.objects import Exactly, datatype, hashable_string_list, string_optional, string_type
 from pants.util.strutil import create_path_env_var
 
 
-class RequirementsPexRequest(datatype([
+class MakePexRequest(datatype([
   ('output_filename', string_type),
   ('requirements', hashable_string_list),
   ('interpreter_constraints', hashable_string_list),
   ('entry_point', string_optional),
+  ('input_files_digest', Exactly(Digest, type(None))),
+  ('source_dirs', hashable_string_list),
```

Comment:
It's unclear how `input_files_digest` and `source_dirs` interact: does the first declare the actual inputs to be included in the workspace, while the latter represents relative paths to capture?

If so, that feels a bit error-prone/redundant. You might consider having the `create_pex` rule get the Snapshot for the `input_files_digest`, and use `snapshot.dirs` to decide what the "source dir" arguments should be...? I'm not 100% certain about the API though.

Reply:
Yeah I think that will work, I'll try it out

Reply:
Actually, I think it might be okay to just hardcode the source dir as `.` (as far as the argument to the pex utility is concerned). That will capture every file that's a child of the tempdir, and since the tempdir is populated from the contents of the digest by `ExecuteProcessRequest` then it means that `pex` will only see whatever files the client of this rule puts into the `Digest` they craft.

Reply:
That _might_ end up capturing the pex binary itself...? Should confirm that it doesn't.

Also, did you know about:
```
--[no-]process-execution-cleanup-local-dirs (default: True)
    Whether or not to cleanup directories used for local process execution
    (primarily useful for e.g. debugging).
```
? Handy for validating the state of things after a process has been executed.

Reply:
Looks like this strategy does capture the pex binary. Doing what you suggested with obtaining the `Snapshot` from a directory would have the problem that files at the top level of `Digest` wouldn't be in any directory, and when we merged that `Digest` with `pex_bin.directory_digest` we either wouldn't be able to pick up top-level .py files, or we'd have to pick up everything including the pex binary. I think what we might want is the ability to merge the request object's `input_files_digest` to a *subdir* of `pex_bin.directory_digest`, but I dont' think we have an intrinsic that will do this right now. Am I missing an easy way to make this happen?

Reply:
> I think what we might want is the ability to merge the request object's input_files_digest to a subdir of pex_bin.directory_digest, but I dont' think we have an intrinsic that will do this right now. Am I missing an easy way to make this happen?

Yea, we don't have that intrinsic currently. It's roughly the inverse of this one though: https://github.com/pantsbuild/pants/blob/6f45bc464d2ad2e46dafcd269f1496dde3953749/src/rust/engine/src/nodes.rs#L267-L294

But note that you can also hack around something like that by executing `mv` as an isolated process ([in bash](https://github.com/pantsbuild/pants/blob/6f45bc464d2ad2e46dafcd269f1496dde3953749/tests/python/pants_test/engine/test_isolated_process.py#L366-L395)): just please open a ticket to reify it if you do.